include/deal.II/base/mpi.h

```diff
@@ -1178,33 +1178,70 @@ namespace Utilities
            const unsigned int root_process = 0);
 
     /**
-     * Sends an object @p object_to_send from the process @p root_process
+     * This function sends an object @p object_to_send from the process @p root_process
      * to all other processes.
      *
-     * A generalization of the classic `MPI_Bcast` function that accepts
-     * arbitrary data types `T`, as long as Utilities::pack() (which in turn
-     * uses `boost::serialize`, see in Utilities::pack() for details) accepts
-     * `T` as an argument.
+     * This function is a generalization of the classic `MPI_Bcast` function
+     * that accepts arbitrary data types `T`, as long as Utilities::pack()
+     * (which in turn uses `boost::serialize`, see in Utilities::pack() for
+     * details) accepts `T` as an argument.
      *
      * @note Be aware that this function is typically a lot more
-     * expensive than an `MPI_Bcast` if you are using simple types
-     * directly supported by MPI. This function will use
+     * expensive than an `MPI_Bcast` because the function will use
      * boost::serialization to (de)serialize, and execute a second
      * `MPI_Bcast` to transmit the size before sending the data
-     * itself. You can also use the other broadcast() function in this
-     * namespace.
+     * itself. On the other hand, if you have a single element of
+     * a data type `T` that is natively supported by MPI, then the
+     * compiler will choose another broadcast() overload that is efficient.
+     * If you have an array of such elements, you should use the other
+     * broadcast() function in this namespace that takes a pointer and a count
+     * argument.
      *
      * @param[in] comm MPI communicator.
      * @param[in] object_to_send An object to send to all processes.
      * @param[in] root_process The process that sends the object to all
      * processes. By default the process with rank 0 is the root process.
      *
+     * @tparam T Any type for which the Utilities::pack() and
+     *   Utilities::unpack() functions can be used to convert the object
+     *   into an array of `char`. The compiler will not select this function
+     *   if `T` is a type that is natively supported by MPI and instead
+     *   use a more efficient overload.
+     *
      * @return On the root process, return a copy of @p object_to_send.
      *   On every other process, return a copy of the object sent by
      *   the @p root_process.
      */
     template <typename T>
-    T
+    typename std::enable_if<is_mpi_type<T> == false, T>::type
+    broadcast(const MPI_Comm &   comm,
+              const T &          object_to_send,
+              const unsigned int root_process = 0);
+
+    /**
+     * This function sends an object @p object_to_send from the process @p root_process
```

Comment:
Does this work well with doxygen? Maybe we should wrap this whole thing in an ``#ifndef DOXYGEN``

Reply:
What do you mean with "this" in "Does this work well with doxygen?"

If we had C++17, it could easily be implemented under the hood with an `if constexpr`, but given what it is, I think it's useful to show both functions in the documentation. Their documentation is different, after all, and I think it's useful to see that the cost of the other function (which you added to the documentation recently) does not apply to this one.

Reply:
I wonder if the two functions will show up correctly in doxygen considering that they only differ by return value. You also already mention the optimization for MPI types in the original function description, so I think it would be enough to only have the main one show up in doxygen.
If you insist to keep it this way, I would like to ask you to check how it renders in doxygen after we merge this.

Reply:
It looks like this: 
![image](https://user-images.githubusercontent.com/7504421/157101038-0f35d168-77c6-470c-b239-ef6d0fd07bdd.png)
This looks good to me.